### DISCUSSION TITLE
[NCR#1 Proposal] TruthOracle: Modular ZK-Enhanced Crypto Prediction Market Resolver using zkEmail

### DISCUSSION THREAD

**Discussion by gittysachin** [noir-lang/noir#5816] [NRG#1 zkEmail]
### **Summary:**
This proposal outlines the development of a modular prediction market platform built on [Berachain](https://docs.berachain.com/learn/), leveraging zkEmail authentication using Noir. Each feature of the platform will be built as a standalone module that can be used as an SDK in other projects. The system will allow users to create prediction markets, deploy smart contracts linked to their email accounts, and verify outcomes using zero-knowledge proofs (ZKPs) derived from exchange data. The modular approach will ensure flexibility, reusability, and integration potential with other projects, while the use of zkEmail will maintain privacy and security.

### **Methodology:**
We will develop an MVP demonstrating the secure and private resolution of crypto prediction markets using zkEmail. The platform will integrate with data sources, generate zero-knowledge proofs for data authenticity, and resolve prediction markets based on verified outcomes. Key steps include porting zk circuits from Circom to Noir, deploying account contract linked with email, developing the token minting process, creating a comprehensive system to handle prediction markets, and reward distribution efficiently.

### **System Design:** 
1. **Modular Architecture**
   - **Objective:** Build the platform with a modular design, where each component (e.g., authentication, market creation, verification) is a separate module.
   - **Approach:** Implement each feature as an independent module that can be plugged into the platform or used as an SDK for other projects. This ensures that the platform can be extended and integrated with ease.

2. **zkEmail Authentication with Noir**
   - **Objective:** Port the zkEmail authentication circuit from Circom to Noir.
   - **Approach:** The zkEmail circuit will be rewritten in Noir ([reference](https://github.com/zkemail/zk-email-verify/blob/main/packages/circuits)), focusing on ensuring privacy-preserving email verification. Users will authenticate their emails when signing up, deploying a smart contract associated with their unique email key.

3. **Wallet: Account Deployment on Sign-up**
   - **Objective:** Link email accounts to smart contracts on the Berachain.
   - **Approach:** When a user signs up, they deposit a small fee to deploy a smart contract account tied to their email. The contract will be responsible for managing the user's prediction market interactions and funds.

4. **Custom Indexer Development**
   - **Objective:** Monitor account events on the Berachain.
   - **Approach:** Develop a custom extendable indexer that listens for market creation, account creation, deposit, withdrawal and prediction invocation events, specifically for zkEmail-authenticated accounts. The indexer will communicate event details to the wallet UI via WebSocket connection, providing real-time updates to users.

5. **Prediction Market Creation and Management**
   - **Objective:** Allow users to create and manage prediction markets.
   - **Approach:** Users can initiate prediction markets for various pairs (e.g., BTC-USDC, ETH-USDT). Markets are community-driven, gaining visibility on the homepage once the betting pool exceeds $500. Each market is defined by a structure with stable collaterals to ensure consistency. The prediction event concludes at the specified time set by the pioneer, typically within a week from the market setup. Community involvement drives market creation and participation.

6. **Oracle Verification with Noir ZKPs**
   - **Objective:** Verify market outcomes using data from multiple exchanges.
   - **Approach:** Implement Noir ZKPs to verify oracle price data received from exchanges like ByBit, Binance, and OKX. This ensures that the outcomes are accurate, secure, and tamper-proof, providing confidence to all participants.

7. **Zero-Knowledge Proof Generation**
    - **Proof Creation:** Generate zk proofs using Noir to verify the authenticity and integrity of the aggregated data from oracles.
    - **Proof Submission:** Submit zk proofs to the smart contract for verification. The smart contract is programmed to verify the correctness of the zk proofs.

8. **Market Resolution**
    - **Outcome Determination:** Once the event concludes, the system uses the verified oracle data to determine the outcome.
    - **Oracle Data Verification:** Retrieve data from oracles (e.g., final price of BTC-USDC), and compare the oracle data against user predictions to determine who was correct.
    - **Proof Verification:** Verify zk proofs to ensure the integrity of the oracle data and the correctness of the outcome. If the zk proofs are valid, the smart contract accepts the proof and updates the state of the prediction market (e.g., resolving the market or updating the outcomes).

9. **Incentive Mechanism**
   - **Objective:** Encourage market creation and participation.
   - **Approach:** Pioneers who initiate markets will receive 1% of the betting pool, while the platform takes an additional 1%. The remaining pool is distributed to winners, ensuring an incentive structure that rewards early adopters and successful participants.

10. **Dashboard**
    - **Objective:** Provide a user interface for interaction and rewards.
    - **Approach:** Develop a dashboard interface where users can create and view markets, track their current predictions, view their earned tokens and prediction history, and manage their account.

### **User Flow:**
1. **User Registration**
   Users securely sign up and authenticate via zkEmail. Their email address is verified using zkEmail, and a smart contract is deployed on the Berachain linked to their email.

2. **Account Deployment**
   Upon registration, users deposit a small fee to deploy a smart contract account on the Berachain, which manages their prediction market interactions and funds.

3. **Prediction Submission**
   Users submit predictions on various crypto-related markets (e.g., BTC-USDC, ETH-USDT) through the platform interface. They can participate in existing markets or create new ones, specifying their predictions and stakes.

4. **Data Collection and Encryption**
   The platform integrates with real-time data sources like exchanges (ByBit, Binance, OKX) to fetch relevant data. This data is collected and encrypted to ensure privacy and security.

5. **Zero-Knowledge Proof Generation**
   zk proofs are generated using Noir to verify the authenticity and integrity of the encrypted data. This ensures that the data used in market resolution is accurate and tamper-proof.

6. **Market Resolution**
   Once the prediction event concludes, the system uses the verified oracle data to determine the outcome. The oracle data is compared against user predictions to identify which predictions are correct.

7. **Outcome Verification**
   zk proofs are used to verify the market outcome, ensuring that the data and predictions are accurate and that the outcome is fair.

8. **Incentive Distribution**
   Rewards are distributed based on the market outcomes. Collaterals are distributed to users with correct predictions, and pioneers who created the market receive their share of the betting pool.

9. **Dashboard Interaction**
   Users interact with the dashboard to create and manage markets, track their current predictions, view earned tokens, and review their prediction history.

### **Tech Stack:**
- **Frontend:** ReactJS + Vite
- **Backend:** NestJS (services - gateway, indexer, relayer, reward, prover)
- **Blockchain:** Berachain (high-performance EVM-identical L1)
- **Smart Contract:** Solidity
- **ZK Proofs:** Noir for zero knowledge circuit development
- **Data Fetching:** ByBit, Binance and OKX oracle price APIs
- **Storage:** On-chain, MongoDB, Redis
- **Communication:** REST, WebSocket, NATS Protocol
- **Containers:** Docker
- **Cloud Compute:** GCP (Google Cloud Platform)

### **Timeline and Deliverables:**

**Week 1: Planning and Setup**
  - Develop detailed system and module designs.
  - Set up the base development environment with initial scalable and reusable modules.

**Week 2: zkEmail Authentication**
- Port zkEmail authentication circuit from Circom to Noir.
- Develop and deploy the smart contract for user account management.

**Week 3: Account and Indexer Development**
- Complete development of the smart contract for user accounts.
- Develop the custom indexer for monitoring account events.
- Integrate with exchange APIs (ByBit, Binance, OKX).

**Week 4: Prediction Market Creation Module**
- Develop the prediction market creation module.
- Complete zkEmail authentication implementation.
- Implement wallet and account deployment processes.

**Week 5: Core Module Development**
- Develop the prediction market management module.
- Implement oracle verification with Noir ZKPs.

**Week 6: zk Proofs and Incentive Mechanism**
- Create zk proof circuits for data integrity.
- Integrate zk proofs with smart contracts for outcome verification.
- Develop and integrate the incentive mechanism for market participants and creators.

**Week 7: Dashboard and Security Features**
- Develop the dashboard interface for user interaction.
- Implement data encryption and secure storage mechanisms.

**Week 8: Testing and Integration**
- Write & conduct unit tests for individual components.
- Execute end-to-end (E2E) tests for the integrated system.
- Complete integration of all modules into a unified system.

**Week 9: Deployment and Documentation**
- Prepare and execute deployment scripts and configurations.
- Deploy the platform to the production environment.
- Develop comprehensive documentation and user guides.

**Week 10: Review and Support**
- Conduct a comprehensive project review and assess outcomes.
- Provide post-launch support and address user feedback.
- Document recommendations for future improvements.
- Outline potential use cases for prediction markets (e.g., sports, stocks, trends, elections, research etc.) and commence initial development for selected use cases.

### **Team:**

- **Sachin Jangid:** Senior Backend Engineer
  - Over 4 years of experience, including 2 years specializing in Web3 and blockchain engineering.
  - Expertise in smart contract development with Cairo and Solidity.
  - Extensive experience in building wallets, indexer, sequencer, order-book and various components of large-scale crypto trading systems.
  - Proficient in TypeScript, NestJS, Rust, and various backend technologies with best engineering practices.
  - Experienced in writing Circom circuits for zkSNARK implementations.
  - Email: [sachinjangid832@gmail.com](mailto:sachinjangid832@gmail.com)
  - GitHub: [gittysachin](https://github.com/gittysachin)
  - LinkedIn: [thesachinjangid](https://www.linkedin.com/in/sachinjangid)

- **Saurav Kumar:** Rockstar Frontend Engineer
  - 4+ years creating and scaling products for multiple startups and entrepreneurs on both front and back-end.
  - More than a year working with EVM based chains on frontend, including wallet integrations and creating custom L3 wallets.
  - Expertise in React, Typescript, NextJS and CD pipelines.
  - Email: [sorvkay@gmail.com](mailto:sorvkay@gmail.com)
  - GitHub: [ksorv](https://github.com/ksorv)

## **Start Date**
- September 9, 2024

**Comment by Divide-By-0** [comment]
Is there a reason you're so interested in berachain? Seems like an odd choice to make when it could be chain agnostic.

**Reply by gittysachin** [reply]
Yes, it is chain agnostic. I wanted to use zk based EVM chain. I felt that Berachain could be used because of its unique proof of liquidity consensus mechanism. It's mainnet launch is around the similar dates of TruthOracle completion date so it's launch might boost the project as well. My another top choice is Scroll. It could be Ethereum itself as well just that the transaction fees are a bit higher.

**Comment by Savio-Sou** [comment]
Selected proposals for NCR#1 are announced: https://github.com/orgs/noir-lang/discussions/5932

Thank you for your support on the initiative. NCR is just one of the many ways to start your explorations with Noir, check the announcement for more potential ways to collaborate.

Personally resonates with the project idea, but was skeptical with the choice of blockchain environment.

Either way, let's bring the world ZK!
